### PR TITLE
HoP can prioritize jobs and leave messages

### PR DESCRIPTION
**Pull request by Super3222**
Refer to https://github.com/yogstation13/yogstation/issues/1651.

### Intent of your Pull Request

This PR will open up the HoP to communicating with potential crewmembers.

1. The HoP will be able to leave a message for new arrivals.
  - Every time the HoP attempts to write or does write a message that information is tracked back to the admins.
  - Every message is logged into the game.

2. The HoP will be able to prioritize jobs.
  - Every job prioritized will show up RED TEXT on the join screen indicating that the station needs that job.
  - The HoP can go back into the list and select that same job to unprioritize it.
  - Again, these actions are logged into the game.

Of course, you've got to have some admin verbs **to regulate this sort of power.**

I have so far coded two.
The first one will reset the new arrival message completely. This will also alert fellow admins.

The second one will LOCK the HoP's power to leave a message. Yes, completely locking them out from doing so. If they try it again, then it will get logged into the game.

*New: And to top it all off - arrival messages will have their own INVESTIGATE logs which admins can instantly pull up without needing to download the entirety of game logs.

I can feel some contemplation for the ability to leave messages. My general thought is that if the HoP exposes the round type, and a player doesn't want to join the round because it's going to be THAT sorta round type then what does it matter? They're going to join, hear the round type over the radio, and then commit suicide. At least in this sense it keeps them off of important jobs. And who's to say they still won't join and suicide?

Thirdly, this PR will also move IsJobAvaliable() from `mob/new_player/` to the jobs subsystem. With this, it can be used universal instead of gauging it on new players.

# **Old** Pictures:

#### Changelog

:cl:
rscadd: The Head of Personnel has the ability to prioritize jobs. Prioritized jobs will show up in red text on the join page.
rscadd: You can only prioritize up to 5 jobs. A general thought is the less amount of jobs you have prioritized the more you'll probably get people to pick them.
experimental The Head of Personnel has the ability to leave a message for new arrivals. These messages are strictly logged and are displayed on the join page.
experimental: The arrival messages are carefully logged into the system. Be careful of what you write.
tweak: IsJobAvaliable() has been moved to the jobs subsystem to be used more universally.
/:cl:

### CODE REVIEW

**Comment by Time-Green** [comment]
GOD BLESS YOU

**Comment by Time-Green** [comment]
Awesome, I like every part of it. 
However I do expect a wave of controversy about leaving messages. 
But I like it at least

**Comment by Dagdammit** [comment]
-I'm worried turning a job red will not be intuitive. A more natural conclusion than "Station's begging for someone to choose this job" would be "This job is somehow an invalid choice right now" (hell players might think they're jobbanned) or "There is some reason to beware this option."
-HoP console and access is perhaps THE most frequently hacked thing, I don't think giving a freshly-gang-converted powergamer assistant the power to shitpost on the "join game" interface itself is the wisest idea. (Giving antags the potential to place their own priority staff requests seems interesting though, so long as they abide by server rules against ganking the new arrivals.)

Personally, I would say:
-Have the message at the top be automated. "Priority Staff Requests: Atmospheric Technician, Captain, Warden." Maybe limit 3 or 4, and maybe have a request automatically end if that job hits its current max headcount.
-Either highlight the roles in question as lightish yellow or just don't highlight them at all.

**Comment by Alek2ander** [comment]
Yeah, message for new arrivals should be on the comms console with captain auth. More secure and makes more sense (it's sending a message to wherever the arrivals shuttle supposedly comes from)

**Comment by Dagdammit** [comment]
Man that console's used by unathorized types even more often, only reason that's not the most frequently hacked machine is that people don't log out after using it. I'm not wholesale opposed to being able to leave a brief message for new arrivals, but it certainly isn't necessary to achiev the whole "we need job X please" intent.

**Comment by Kmc2000** [comment]
Change from red, red heavily implies that it's banned or something, like when i first saw this that's what it looked like

**Comment by Stealthkibbler** [comment]
Should change it from being red to being highlighted yellow with an exclamation mark

File: code/controllers/subsystem/jobs.dm [hunk 1]
```diff
@@ -492,4 +492,28 @@ var/datum/subsystem/job/SSjob
 					return
 				newjob.total_positions = J.total_positions
 				newjob.spawn_positions = J.spawn_positions
-				newjob.current_positions = J.current_positions
\ No newline at end of file
+				newjob.current_positions = J.current_positions
+
+// Use it for generic or non-generic (in new_player.dm) reasons.
+/datum/subsystem/job/proc/IsJobAvailable(rank, var/mob/new_player/NP)
```
Comment (X-TheDark) [inline code/controllers/subsystem/jobs.dm:499]: Would be nice to lose the var and just `mob/new_player/NP` it.

**Comment by X-TheDark** [comment]
Very nice, great job! Minor gripe about how badly the red contrasts, though. Not sure what other color would look good.

**Comment by Slinky5890** [comment]
I feel like green would be a good colour. Red or yellow to me does shout 'warning', like maybe the role is now full or is almost full

**Comment by Qbopper** [comment]
It needs to be made clearer what the highlighted jobs mean - they just look like they're discouraging you from selecting it in that miage, there's no context for any players unaware of the PR

**Comment by DoctorPork** [comment]
Make it greentext green and bold instead of red.

File: code/controllers/subsystem/jobs.dm (hunk 1 above)
Comment (Super3222) [reply code/controllers/subsystem/jobs.dm:499]: I did it like that and it gave me an undefined variable error. These things are probably defined in a really minuscule pathed object / datum, before used without the var/ in much longer pathed children.

**Comment by Super3222** [comment]
I'll take everyone's comments into consideration and update this. Thanks guys! My next commit will hopefully make the intent more clearer, and change the color.

EDIT: As for moving the arrivals message to the communications console, I only wish I could. But there's a few problems with that. For one, the Captain is going to be messing with it all the time. That takes it away from the HoP. The next commit will create an ID lock on both prioritizing and the arrivals message.

**Comment by harmonyn** [comment]
tgstation here, this looks p. cool and I am porting it with modifications

also >all these trialmins

**Comment by Super3222** [comment]
Aw man. I forgot to save the new screenshots!

File: code/game/machinery/computer/card.dm [hunk 2]
```diff
@@ -472,6 +477,53 @@ var/time_last_changed_position = 0
...
+				return
+			var/mob/M = usr
+			if(ticker.id_console_msg_lock)
+				M << "<span class='warning'>Central Command has blocked your station's power to do this.</span>"
+				log_game("[M] ([M.ckey]) has attempted to change the arrivals message while it was locked.")
+				return
+
+			message_admins("[M.name] ([M.ckey]) is changing the arrival message.")
+			log_game("[M] ([M.ckey]) is attempting to change the arrivals message ([ticker.identification_console_message]).")
+			var/msg = input(usr, "What do you want centcomm to tell potential employees enlisting to [station_name()]?", "Arrivals Message")
```
Comment (Cruix) [inline code/game/machinery/computer/card.dm]: You need to check if they still have ID access and are still adjacent to the console

File: code/game/machinery/computer/card.dm [hunk 3]
```diff
@@ -472,6 +477,53 @@ var/time_last_changed_position = 0
...
+				return
+			var/mob/M = usr
+			var/list/jobs = list()
+			for(var/datum/job/job in SSjob.occupations)
+				if(job && SSjob.IsJobAvailable(job.title))
+					if(job.title != "Assistant")
+						jobs += job.title
+
+			if(length(jobs))
+				var/pickjob = input(usr,"Select Job","Prioritize/Un-Prioritize",null) as anything in jobs
```
Comment (Cruix) [inline code/game/machinery/computer/card.dm:523]: You need to check for canusetopic() and id access again

File: code/game/machinery/computer/card.dm (hunk 2 above)
Comment (Cruix) [inline code/game/machinery/computer/card.dm:496]: You need to check if they still have ID access and pass canUseTopic()

File: code/game/machinery/computer/card.dm [hunk 4]
```diff
@@ -472,6 +477,53 @@ var/time_last_changed_position = 0
...
+			if(ticker.id_console_msg_lock)
+				M << "<span class='warning'>Central Command has blocked your station's power to do this.</span>"
+				log_game("[M] ([M.ckey]) has attempted to change the arrivals message while it was locked.")
+				return
+
+			message_admins("[M.name] ([M.ckey]) is changing the arrival message.")
+			log_game("[M] ([M.ckey]) is attempting to change the arrivals message ([ticker.identification_console_message]).")
+			var/msg = input(usr, "What do you want centcomm to tell potential employees enlisting to [station_name()]?", "Arrivals Message")
+			if(msg)
+				message_admins("[M.name] ([M.ckey]) has changed the arrivals message to: [msg]")
```
Comment (Cruix) [inline code/game/machinery/computer/card.dm]: This should have an href to reset it or show the name of the verb that resets it

File: code/modules/admin/verbs/identifcation_console.dm [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/client/proc/reset_idconsole_msg()
+	set category = "Special Verbs"
+	set name = "resetidconsole"
```
Comment (Cruix) [inline code/modules/admin/verbs/identifcation_console.dm]: Give this a proper name like all the other admin verbs

File: code/modules/admin/verbs/identifcation_console.dm [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	if(!check_rights(0))
+		return
+
+	message_admins("[usr.ckey] has reset the identification console message. Previous message: [ticker.identification_console_message]")
+	log_admin("[usr.ckey] has reset the identification console message. Previous message: [ticker.identification_console_message]")
+	ticker.identification_console_message = null
+
+/client/proc/lock_idconsole()
+	set category = "Special Verbs"
+	set name = "toggle_idconsolelock"
```
Comment (Cruix) [inline code/modules/admin/verbs/identifcation_console.dm]: Give this a proper name as well

File: code/game/machinery/computer/card.dm (hunk 4 above)
Comment (Super3222) [reply code/game/machinery/computer/card.dm]: Nice thinking.

File: code/game/machinery/computer/card.dm [hunk 7]
```diff
@@ -472,6 +477,59 @@ var/time_last_changed_position = 0
...
+			if(ticker.id_console_msg_lock)
+				M << "<span class='warning'>Central Command has blocked your station's power to do this.</span>"
+				log_game("[M] ([M.ckey]) has attempted to change the arrivals message while it was locked.")
+				return
+
+			message_admins("[M.name] ([M.ckey]) is changing the arrival message.")
+			log_game("[M] ([M.ckey]) is attempting to change the arrivals message ([ticker.identification_console_message]).")
+			if(!allowed(usr) || !usr.canUseTopic(src,be_close=TRUE))
+				message_admins("[M.name] ([M.ckey] (ckey)) did not have the correct ID, or simply was interrupted when trying to change the arrival message.")
+				return
```
Comment (Cruix) [inline code/game/machinery/computer/card.dm]: This needs to go under the input

File: code/game/machinery/computer/card.dm [hunk 8]
```diff
@@ -472,6 +477,59 @@ var/time_last_changed_position = 0
...
+			var/mob/M = usr
+			var/list/jobs = list()
+			for(var/datum/job/job in SSjob.occupations)
+				if(job && SSjob.IsJobAvailable(job.title))
+					if(job.title != "Assistant")
+						jobs += job.title
+
+			if(length(jobs))
+				if(!M.canUseTopic(src,be_close=TRUE))
+					return
```
Comment (Cruix) [inline code/game/machinery/computer/card.dm]: This also needs to go under the input

**Comment by Super3222** [comment]
In the last commit I cap the amount of jobs the HoP can prioritize to 5.

File: code/controllers/subsystem/jobs.dm (hunk 1 above)
Comment (Super3222) [reply code/controllers/subsystem/jobs.dm:499]: Nevermind, I will change it to that as soon as I can.

**Pull request merged**